agent/fw_drupal.c

```diff
@@ -155,40 +227,108 @@ NR_PHP_WRAPPER(nr_drupal_http_request_exec) {
    * (nr_php_swap_user_functions), which breaks as since PHP 7.3 user
    * functions are stored in shared memory.
    */
-  bool is_drupal_7;
-  zval* arg = NULL;
-  zend_execute_data* ex = nr_get_zend_execute_data(NR_EXECUTE_ORIG_ARGS);
+  nr_drupal_http_request_add_headers(NR_EXECUTE_ORIG_ARGS);
 
-  if (NULL == ex) {
-    NR_PHP_WRAPPER_LEAVE
-  }
+  NR_PHP_WRAPPER_REQUIRE_FRAMEWORK(NR_FW_DRUPAL);
 
-  is_drupal_7 = (ex->func->common.num_args == 2);
+  NRPRG(drupal_http_request_depth) += 1;
 
   /*
-   * If only one argument is given, an empty list is inserted as second
-   * argument. NR headers are added during a later step.
+   * We only want to create a metric here if this isn't a recursive call to
+   * drupal_http_request() caused by the original call returning a redirect.
+   * We can check how many drupal_http_request() calls are on the stack by
+   * checking a counter.
    */
-  if (ZEND_NUM_ARGS() == 1) {
-    arg = nr_php_zval_alloc();
-    array_init(arg);
+  if (1 == NRPRG(drupal_http_request_depth)) {
+    NRPRG(drupal_http_request_segment)
+        = nr_segment_start(NRPRG(txn), NULL, NULL);
+  }
+}
+NR_PHP_WRAPPER_END
 
-    nr_php_arg_add(NR_EXECUTE_ORIG_ARGS, arg);
+NR_PHP_WRAPPER(nr_drupal_http_request_after) {
+  zval* arg1 = NULL;
 
-    nr_php_zval_free(&arg);
+  (void)wraprec;
+
+  NR_PHP_WRAPPER_REQUIRE_FRAMEWORK(NR_FW_DRUPAL);
+
+  /*
+   * Grab the URL for the external metric, which is the first parameter in all
+   * versions of Drupal.
+   */
+  arg1 = nr_php_arg_get(1, NR_EXECUTE_ORIG_ARGS TSRMLS_CC);
+  if (0 == nr_php_is_zval_non_empty_string(arg1)) {
+    goto end;
   }
 
   /*
-   * nr_php_get_user_func_arg is used, as nr_php_arg_get calls ZVAL_DUP
-   * on the argument zval and thus doesn't allow us to change the
-   * original argument.
+   * We only want to create a metric here if this isn't a recursive call to
```

Comment:
What happens if it is a recursive call based on a redirect?

Reply:
The global `NRPRG(drupal_http_request_segment)` segment remains unchanged if it is a recursive call based on a redirect. `NRPRG(drupal_http_request_segment)` segment is only created by the first call to `drupal_http_request`  - see [agent/fw_drupal.c:242](https://github.com/misiekn/newrelic-php-agent/blob/oapi/drupal_http_request/agent/fw_drupal.c#L242). It is only destroyed when the first call to `drupal_http_request` returns - see [agent/fw_drupal.c:291](https://github.com/misiekn/newrelic-php-agent/blob/oapi/drupal_http_request/agent/fw_drupal.c#L291). It was better visible in pre-oapi instrumentation, because everything happened within the scope of a single function - `drupal_http_request_exec` - and the segment was local to that function. There was no need to maintain state before and after call to `drupal_http_request`. With oapi the management of external segment must be split between before and after callbacks and it's easier to lose track of what is going on. For now `drupal_http_request` oapi instrumentation uses a simple solution of maintaining external segment as a global variable. This will probably have to be revisited when fibers support is added to the agent.

Reply:
Got it, thanks for the explanation :)